### ISSUE TITLE
tensorpack/examples/A3C-Gym/train-atari.py

### ISSUE THREAD

**Issue by lezhang-thu** [ppwwyyxx/tensorpack#358]
`value_loss = tf.nn.l2_loss(self.value - futurereward, name='value_loss')`
Why not write it as `value_loss = tf.reduce_sum(self.value * advantage, name='value_loss')`?

The latter seems simpler and consistent with policy loss if it were correct.
Thanks!

**Comment by ppwwyyxx** [comment]
I think what you meant to say is `value_loss = tf.reduce_sum(advantage * advantage, name='value_loss'`.

And this is wrong, because advantage has `stop_gradient`.

**Comment by lezhang-thu** [comment]
Thanks for the instant reply.

What I mean is what is the gradient of `tf.nn.l2_loss`?
The answer seems to be just the `advantage * \nabla self.value`.
So  one really needs the `stop_gradient` of `advantage` to make the gradient be correct.

I'm not sure of its correctness. So I'm just asking you for the advice.

**Comment by ppwwyyxx** [comment]
Ah I see what you are talking about. The two equation actually have different value, but you're right they have the same gradient.

I think it's fine to keep it in the current state to be consistent with the formulation used in the paper.

**Comment by lezhang-thu** [comment]
Thanks! Now I am sure of it.

**Comment by lezhang-thu** [comment]
EVALUATE_PROC in train-atari.py is of no use. Maybe it's a good idea to remove it.
